create-next-app: pnpm v10 blocks sharp build

### Link to the code that reproduces this issue

https://github.com/karlhorky/repro-create-next-app-pnpm-ignored-build-scripts

### To Reproduce

1. Use pnpm v10+
2. Use `create-next-app` to create an app (flags are irrelevant, occurs with any flags), eg:
   ```bash
   $ pnpm create next-app@15.5.1-canary.14 . --app --no-src-dir --no-eslint --import-alias @/\* --no-tailwind --typescript --turbopack
   ```
3. Observe a pnpm warning (see below) that the `sharp` build script has not been run, with a message to run `pnpm approve-builds`
4. Optional: configure [`strict-dep-builds=true` in `~/.npmrc`](https://github.com/pnpm/pnpm/pull/9071) and the warning becomes an error, with a non-zero exit code

Warning:

```bash
$  pnpm create next-app@15.5.1-canary.14 . --app --no-src-dir --no-eslint --import-alias @/\* --no-tailwind --typescript --turbopack
Creating a new Next.js app in /Users/k/p/b.

Using pnpm.

Initializing project with template: app


Installing dependencies:
- react
- react-dom
- next

Installing devDependencies:
- typescript
- @types/node
- @types/react
- @types/react-dom

Packages: +32
++++++++++++++++++++++++++++++++
Progress: resolved 60, reused 33, downloaded 0, added 32, done

dependencies:
+ next 15.5.1-canary.14 (15.5.2 is available)
+ react 19.1.0 (19.1.1 is available)
+ react-dom 19.1.0 (19.1.1 is available)

devDependencies:
+ @types/node 20.19.11 (24.3.0 is available)
+ @types/react 19.1.12
+ @types/react-dom 19.1.8
+ typescript 5.9.2

╭ Warning ───────────────────────────────────────────────────────────────────────────────────╮
│                                                                                            │
│   Ignored build scripts: sharp.                                                            │
│   Run "pnpm approve-builds" to pick which dependencies should be allowed to run scripts.   │
│                                                                                            │
╰────────────────────────────────────────────────────────────────────────────────────────────╯

Done in 1.8s using pnpm v10.15.0
Initialized a git repository.

Success! Created b at /Users/k/p/b
```

Error with `strict-dep-builds=true` in `~/.npmrc`:

```bash
$ pnpm create next-app@15.5.1-canary.14 . --app --no-src-dir --no-eslint --import-alias @/\* --no-tailwind --typescript --turbopack

Creating a new Next.js app in /Users/k/p/a.

Using pnpm.

Initializing project with template: app


Installing dependencies:
- react
- react-dom
- next

Installing devDependencies:
- typescript
- @types/node
- @types/react
- @types/react-dom

Packages: +32
++++++++++++++++++++++++++++++++
Downloading next@15.5.1-canary.14: 30.35 MB/30.35 MB, done
Downloading @next/swc-darwin-arm64@15.5.1-canary.14: 42.72 MB/42.72 MB, done
Progress: resolved 60, reused 27, downloaded 6, added 32, done

dependencies:
+ next 15.5.1-canary.14 (15.5.2 is available)
+ react 19.1.0 (19.1.1 is available)
+ react-dom 19.1.0 (19.1.1 is available)

devDependencies:
+ @types/node 20.19.11 (24.3.0 is available)
+ @types/react 19.1.12
+ @types/react-dom 19.1.8
+ typescript 5.9.2

 ERR_PNPM_IGNORED_BUILDS  Ignored build scripts: sharp

Run "pnpm approve-builds" to pick which dependencies should be allowed to run scripts.

Aborting installation.
  pnpm install has failed.
```

### Current vs. Expected behavior

Current: pnpm warning or error on `create-next-app`

Expected: no pnpm warning or error on `create-next-app`

### Suggested Solution

Configure `pnpm.onlyBuiltDependencies` with `'sharp'` in the generated `package.json` as mentioned in the pnpm v10 release notes:

- https://github.com/pnpm/pnpm/releases/tag/v10.0.0

`package.json`

```json
{
  "pnpm": {
    "onlyBuiltDependencies": ["sharp"]
  }
}
```

Alternatively, `onlyBuiltDependencies` can also be set in [a `pnpm-workspace.yaml` file](https://pnpm.io/settings#onlybuiltdependencies). But currently `pnpm-workspace.yaml` is not generated by `create-next-app`.

```yaml
onlyBuiltDependencies:
- sharp
```

### Provide environment information

```bash
Operating System:
  Platform: darwin
  Arch: arm64
  Version: Darwin Kernel Version 24.6.0: Mon Jul 14 11:30:34 PDT 2025; root:xnu-11417.140.69~1/RELEASE_ARM64_T8103
  Available memory (MB): 16384
  Available CPU cores: 8
Binaries:
  Node: 22.18.0
  npm: 10.9.3
  Yarn: 1.22.22
  pnpm: 10.15.0
Relevant Packages:
  next: 15.5.1-canary.14 // Latest available version is detected (15.5.1-canary.14).
  eslint-config-next: N/A
  react: 19.1.0
  react-dom: 19.1.0
  typescript: 5.9.2
Next.js Config:
  output: N/A
```

### Which area(s) are affected? (Select all that apply)

create-next-app

### Which stage(s) are affected? (Select all that apply)

next start (local), next build (local), next dev (local)

### Additional context

--

> I guess, since the CLI knows what package manager is being used (right?), it could append this to the `package.json` for `pnpm`.

Yeah, that seems to make sense to me. It appears `create-next-app` already sets some parts conditionally in the `package.json`:

https://github.com/vercel/next.js/blob/69af581db3095c9ef5636e6296395b303fbf1fdf/packages/create-next-app/templates/index.ts#L187-L208

> do you know what are the implications of the post install for sharp not running?

If build scripts for packages are not run, the package may be unusable:

```bash
$ mkdir a && cd a
$ pnpm init
...
$ pnpm add bcrypt@5
...
$ echo "import bcrypt from 'bcrypt'; bcrypt.hash('hello', 10).then(console.log);" > index.ts
$ node index.ts
node:internal/modules/cjs/loader:1365
  const err = new Error(message);
              ^

Error: Cannot find module '/Users/k/p/a/node_modules/.pnpm/bcrypt@5.1.1/node_modules/bcrypt/lib/binding/napi-v3/bcrypt_lib.node'
Require stack:
- /Users/k/p/a/node_modules/.pnpm/bcrypt@5.1.1/node_modules/bcrypt/bcrypt.js
    at Function._resolveFilename (node:internal/modules/cjs/loader:1365:15)
    at defaultResolveImpl (node:internal/modules/cjs/loader:1021:19)
    at resolveForCJSWithHooks (node:internal/modules/cjs/loader:1026:22)
    at Function._load (node:internal/modules/cjs/loader:1175:37)
    at TracingChannel.traceSync (node:diagnostics_channel:322:14)
    at wrapModuleLoad (node:internal/modules/cjs/loader:235:24)
    at Module.require (node:internal/modules/cjs/loader:1445:12)
    at require (node:internal/modules/helpers:135:16)
    at Object.<anonymous> (/Users/k/p/a/node_modules/.pnpm/bcrypt@5.1.1/node_modules/bcrypt/bcrypt.js:6:16)
    at Module._compile (node:internal/modules/cjs/loader:1688:14) {
  code: 'MODULE_NOT_FOUND',
  requireStack: [
    '/Users/k/p/a/node_modules/.pnpm/bcrypt@5.1.1/node_modules/bcrypt/bcrypt.js'
  ]
}

Node.js v22.18.0
```

## Workaround (until PR #83168 merged)

Use [`pnpm add next --allow-build=sharp`](https://pnpm.io/cli/add#--allow-build) and [`pnpm rebuild sharp`](https://pnpm.io/cli/rebuild) as extra commands after `create-next-app` to explicitly approve `sharp` non-interactively without any other changes to the project.

Example:

```bash
$ pnpm create next-app . --app --no-turbopack --no-src-dir --no-eslint --import-alias @/\* --no-tailwind --typescript
Creating a new Next.js app in /Users/k/p/repro-conflicting-dependabot-and-renovate-config-vs-pnpm-minimum-release-age.

Using pnpm.

Initializing project with template: app


Installing dependencies:
- react
- react-dom
- next

Installing devDependencies:
- typescript
- @types/node
- @types/react
- @types/react-dom

Packages: +35
+++++++++++++++++++++++++++++++++++
Downloading @img/sharp-libvips-darwin-arm64@1.0.4: 7.10 MB/7.10 MB, done
Downloading next@15.1.7: 25.16 MB/25.16 MB, done
Downloading @next/swc-darwin-arm64@15.1.7: 41.27 MB/41.27 MB, done
Progress: resolved 60, reused 13, downloaded 23, added 35, done

dependencies:
+ next 15.1.7 (15.5.3 is available)
+ react 19.1.1
+ react-dom 19.1.1

devDependencies:
+ @types/node 20.19.15 (24.5.2 is available)
+ @types/react 19.1.13
+ @types/react-dom 19.1.9
+ typescript 5.9.2

 ERR_PNPM_IGNORED_BUILDS  Ignored build scripts: sharp

Run "pnpm approve-builds" to pick which dependencies should be allowed to run scripts.

Aborting installation.
  pnpm install has failed.

A new version of `create-next-app` is available!
You can update by running: pnpm add -g create-next-app

$ cat package.json
{
  "name": "repro-conflicting-dependabot-and-renovate-config-vs-pnpm-minimum-release-age",
  "version": "0.1.0",
  "private": true,
  "scripts": {
    "dev": "next dev",
    "build": "next build",
    "start": "next start",
    "lint": "next lint"
  },
  "dependencies": {
    "next": "15.1.7",
    "react": "^19.0.0",
    "react-dom": "^19.0.0"
  },
  "devDependencies": {
    "@types/node": "^20",
    "@types/react": "^19",
    "@types/react-dom": "^19",
    "typescript": "^5"
  }
}

$ ls -al pnpm-workspace.yaml
ls: pnpm-workspace.yaml: No such file or directory

$ pnpm add next --allow-build=sharp
Already up to date
Progress: resolved 60, reused 36, downloaded 0, added 0, done
Done in 451ms using pnpm v10.17.1

$ pnpm rebuild sharp
node_modules/.pnpm/sharp@0.33.5/node_modules/sharp: Running install script, done in 1.6s

$ # No change to package.json
$ cat package.json
{
  "name": "repro-conflicting-dependabot-and-renovate-config-vs-pnpm-minimum-release-age",
  "version": "0.1.0",
  "private": true,
  "scripts": {
    "dev": "next dev",
    "build": "next build",
    "start": "next start",
    "lint": "next lint"
  },
  "dependencies": {
    "next": "15.1.7",
    "react": "^19.0.0",
    "react-dom": "^19.0.0"
  },
  "devDependencies": {
    "@types/node": "^20",
    "@types/react": "^19",
    "@types/react-dom": "^19",
    "typescript": "^5"
  }
}

$ # But the `pnpm-workspace.yaml` file has been created, allowing builds of `sharp`
$ cat pnpm-workspace.yaml
onlyBuiltDependencies:
  - sharp
```

--- a/packages/create-next-app/templates/index.ts
+++ b/packages/create-next-app/templates/index.ts
@@ -320,14 +320,19 @@ export const installTemplate = async ({
   }
 
   if (packageManager === "pnpm") {
-    // TODO: This is a v9 setting that still works in v10. Use pnpm-workspace.yaml
-    // when v10 is used
-    packageJson.pnpm = {
-      // Sharp has prebuilt binaries for the platforms next-swc has binaries.
-      // If it needs to build binaries from source, next-swc wouldn't work either.
-      // See https://sharp.pixelplumbing.com/install/#:~:text=When%20using%20pnpm%2C%20add%20sharp%20to%20ignoredBuiltDependencies%20to%20silence%20warnings
-      neverBuiltDependencies: ["sharp"],
-    };
+    const pnpmWorkspaceYaml = [
+      // required for v9, v10 doesn't need it anymore
+      "packages:",
+      "  - .",
+      // v10 setting without counterpart in v9
+      "ignoredBuiltDependencies:",
+      "  - sharp",
+      "",
+    ].join(os.EOL);
+    await fs.writeFile(
+      path.join(root, "pnpm-workspace.yaml"),
+      pnpmWorkspaceYaml,
+    );
   }
 
   await fs.writeFile(

--- a/packages/create-next-app/templates/index.ts
+++ b/packages/create-next-app/templates/index.ts
@@ -319,10 +319,14 @@ export const installTemplate = async ({
     packageJson.devDependencies = sorted(packageJson.devDependencies);
   }
 
-  /* For pnpm v10+ x-ref: https://github.com/vercel/next.js/issues/83158 */
   if (packageManager === "pnpm") {
+    // TODO: This is a v9 setting that still works in v10. Use pnpm-workspace.yaml
+    // when v10 is used
     packageJson.pnpm = {
-      onlyBuiltDependencies: ["sharp"],
+      // Sharp has prebuilt binaries for the platforms next-swc has binaries.
+      // If it needs to build binaries from source, next-swc wouldn't work either.
+      // See https://sharp.pixelplumbing.com/install/#:~:text=When%20using%20pnpm%2C%20add%20sharp%20to%20ignoredBuiltDependencies%20to%20silence%20warnings
+      neverBuiltDependencies: ["sharp"],
     };
   }
 

--- a/packages/create-next-app/templates/index.ts
+++ b/packages/create-next-app/templates/index.ts
@@ -319,6 +319,13 @@ export const installTemplate = async ({
     packageJson.devDependencies = sorted(packageJson.devDependencies);
   }
 
+  /* For pnpm v10+ x-ref: https://github.com/vercel/next.js/issues/83158 */
+  if (packageManager === "pnpm") {
+    packageJson.pnpm = {
+      onlyBuiltDependencies: ["sharp"],
+    };
+  }
+
   await fs.writeFile(
     path.join(root, "package.json"),
     JSON.stringify(packageJson, null, 2) + os.EOL,

--- a/packages/create-next-app/templates/index.ts
+++ b/packages/create-next-app/templates/index.ts
@@ -326,7 +326,12 @@ export const installTemplate = async ({
       "  - .",
       // v10 setting without counterpart in v9
       "ignoredBuiltDependencies:",
+      // Sharp has prebuilt binaries for the platforms next-swc has binaries.
+      // If it needs to build binaries from source, next-swc wouldn't work either.
+      // See https://sharp.pixelplumbing.com/install/#:~:text=When%20using%20pnpm%2C%20add%20sharp%20to%20ignoredBuiltDependencies%20to%20silence%20warnings
       "  - sharp",
+      // Not needed for pnpm: https://github.com/unrs/unrs-resolver/issues/193#issuecomment-3295510146
+      "  - unrs-resolver",
       "",
     ].join(os.EOL);
     await fs.writeFile(

